src/tarantool.client/LogicalConnection.cs

```diff
@@ -20,20 +20,73 @@ internal class LogicalConnection : ILogicalConnection
     {
         private readonly MsgPackContext _msgPackContext;
 
-        private readonly INetworkStreamPhysicalConnection _physicalConnection;
+        private readonly ClientOptions _clientOptions;
 
-        private long _currentRequestId;
+        private readonly RequestIdCounter _requestIdCounter;
 
-        private readonly ConcurrentDictionary<RequestId, TaskCompletionSource<MemoryStream>> _pendingRequests =
-            new ConcurrentDictionary<RequestId, TaskCompletionSource<MemoryStream>>();
+        private INetworkStreamPhysicalConnection _physicalConnection;
+
+        private IResponseReader _responseReader;
 
         private readonly ILog _logWriter;
 
-        public LogicalConnection(ClientOptions options, INetworkStreamPhysicalConnection physicalConnection)
+        public LogicalConnection(ClientOptions options, RequestIdCounter requestIdCounter)
         {
+            _clientOptions = options;
+            _requestIdCounter = requestIdCounter;
             _msgPackContext = options.MsgPackContext;
             _logWriter = options.LogWriter;
-            _physicalConnection = physicalConnection;
+        }
+
+        public void Dispose()
+        {
+            Interlocked.Exchange(ref _responseReader, null)?.Dispose();
+            Interlocked.Exchange(ref _physicalConnection, null)?.Dispose();
+        }
+
+        public async Task Connect()
+        {
+            _physicalConnection = new NetworkStreamPhysicalConnection();
+            await _physicalConnection.Connect(_clientOptions);
+
+            var greetingsResponseBytes = new byte[128];
+            var readCount = await _physicalConnection.ReadAsync(greetingsResponseBytes, 0, greetingsResponseBytes.Length);
+            if (readCount != greetingsResponseBytes.Length)
+            {
+                throw ExceptionHelper.UnexpectedGreetingBytesCount(readCount);
+            }
+
+            var greetings = new GreetingsResponse(greetingsResponseBytes);
+
+            _clientOptions.LogWriter?.WriteLine($"Greetings received, salt is {Convert.ToBase64String(greetings.Salt)} .");
+
+            _responseReader = new ResponseReader(_clientOptions, _physicalConnection);
+            _responseReader.BeginReading();
+
+            _clientOptions.LogWriter?.WriteLine("Server responses reading started.");
+
+            await LoginIfNotGuest(greetings);
+        }
+
+        public bool IsConnected()
+        {
+            return !(_responseReader?.IsFaultedState ?? true) && (_physicalConnection?.IsConnected() ?? false);
```

Comment:
Initialization of ResponseReader and PhysicalConnection should be moved to constructor. Then all fields could be made readonly and there will be no reason for testing them for null.